### ISSUE TITLE
Inconsistent results when installing a wheel with multiple packages using setuptools and pip

### ISSUE THREAD

**Issue by wvxvw** [pypa/packaging-problems#227]
So, it is possible to specify multiple packages in `setup(packages=[...])`.  The problem... a wheel that results from such a creature will, indeed contain several packages, but when this wheel is depended on by another package, and `setup(install_requires=['package_from_wheel_with_multiple_packages'])` tries to install it all sorts of wrong things begin to happen, but nothing right happens.  The resulting wheel will only reflect the name of one of the packages that reside in it, and, ultimately, that's what's going to be installed (I think).  This is while `pip` extracts everything from the wheel and copies it into `site-packages`, whatever it is, so you end up installing multiple packages, while, technically, you still cannot specify that you depend on any package other than the one which was chosen to be used as the wheel's name...

----

Bottom line: I suggest that you disallow specifying multiple packages inside a single wheel.  And if my dreams will come true before I retire from this job, maybe, one day, you can create something like jumbo-jar, let's call it a mega-wheel, for more grandiose effect, where you can package several wheels together, which will not be used in dependency resolution, but only as a deployment vehicle.

**Comment by ncoghlan** [comment]
cloudify's wagon format is also handy if you just want a way to bundle up multiple wheel archive for subsequent installation in a way that doesn't require internet access: https://github.com/cloudify-cosmo/wagon

We could potentially do more to highlight that format, so if there's an addition to the Python Packaging User Guide that you think may have helped you find that project (assuming it's what you're looking for), that could be a good way of resolving your question.

**Comment by wvxvw** [comment]
@jaraco No, I'm not confused.  But there's also no contradiction between your explanation and what I wrote, so I don't really know what to tell you...

One thing where we disagree is in whether the feature works (well), and whether that's a good idea to try to use it.  In my experience, the feature mostly doesn't work, and so should be either removed or fixed or avoided on the part of the end user.  I keep running into this feature not working with a bunch of packages, like `azure` and `azure-common`, all sorts of packages provided by Google for its GCE services and so on.  This happens in particular when people are trying to get smart and use namespace packages (which is another seldom working feature of Python packaging), or when they try to make packages that install packages with the name different than the one appearing on the wheel or source distribution.

So, for example, try making a vanilla `setup.py` which installs both `azure` and `azure-common` and see what a cluster pain you will be in after trying to install this.  Try both `pip` install` it and `setup.py install` it, also try installing it from another package that depends on it: there are too many things going wrong at the same time.  So, the first passage you quoted is hardly a hyperbola.

And don't make me started on what happens when packages supply data files.  The entire Python packaging is like walking on egg shells, but if it gets to distributing something that's not Python source... nothing works, as in all results are inconsistent and unpredictable.  It's a miracle that people don't complain about it that much.

----

@merwok `zipapp` is a laughing stock... I mean, it's immediate in the name: who on earth would package a distribution that you need to install into a *Zip* archive...  You want to package a file-system, not just files' contents. Else you have to handle file's permissions and users on your own... but this is only the tip of the iceberg.  Digging further into Python packaging I realized that binary installs are basically not usable... it's surprising that so many people would use them, even in production, while the whole system is fundamentally dysfunctional.

----

@ncoghlan I'll need to research this. Thank you for the link.  But, I came to conclusion that Wheel format is fundamentally defective and inappropriate for packaging and distribution of software in general. I might still be happier if I could piggyback on someone else's work when it comes to packaging, but knowing that it uses Wheel inside will cause me unrest and a strong desire to redo it.

**Comment by gaborbernat** [comment]
You mentioned multiple times how things cause all kinds of problems. Instead of trying to sand bag the format, can you please submit bug reports of such problems so we can fix those. Wheel format although not perfect works well for a lot of people in my experience, granted it may have some shortcomings/bugs, but let's address those via those bug/feature reports.

**Comment by wvxvw** [comment]
> can you please submit bug reports of such problems so we can fix those.

That's a lot of work... plus, I did submit some.  I'll submit some more, but I don't believe you will be able to *fix* them.  The proper solution is to *redo packaging from scratch*.  All existing solutions are light years away from a state of decent usability and they are not improving over time...

----

@ncoghlan OMG, I just opened the sources... no.  This is so bad, I don't even know where to begin to explain how bad it is.

**Comment by pfmoore** [comment]
> The proper solution is to redo packaging from scratch

Please let us know here when you've completed that project ;-)

> I did submit some.

If you can provide pointers to them, I can look and see if there's anything I can do on them. But you'd need to be a lot clearer on details of what you're trying to do, and why, and get rid of all the rhetoric, for them to be actionable - so if they have the same tone as you've taken here, I doubt there will be much I (or anyone else) can do.

**Comment by wvxvw** [comment]
In many cases I run into answers implying that "the problem doesn't exist" because it's broken by design.  Like, for example, if you don't specify in `setup.py` that `zip_safe=False`, then most of `pkg_resources` code doesn't work (because it tries to extract directories from a Zip archive, which, obviously doens't have them).  But when I'm pointing this out, I'm told that it works as expected...

But then you have packages that you didn't package, and you have no control over their metadata... and then it basically means that `pkg_resources` is a kind of timer bomb waiting to explode eventually.  But nobody cares.

Or when I say that specifying non-Python source files in MANIFEST doesn't work because you cannot do this conditionally, as in, per platform or per architecture, I'm told that "in practice that is not an issue", whereas what I need to have done is not considered a "real practice", while the imaginary practice of the person who invented the idiotic format is consider the "true and real practice"...

I just really need to get back to work, otherwise I could extend this list with dozens of entries.

**Comment by njsmith** [comment]
I'm sorry you're unhappy. Packaging does have issues, and is often confusing.

But your way of expressing that unhappiness is to shout incoherently at a bunch of volunteers who are patiently trying to figure out what your problems are and how to help, and that's not ok. We are not your punching bag.

You need to either calm down, apologize, and start listening and engaging productively, or you need to go vent somewhere else.

**Comment by wvxvw** [comment]
@njsmith I'm equally sorry that you are so confused about the nature of this project.  Unfortunately, I as well as a lot of other programmers, are held hostages of projects like this one.  Even though I manage a department in the company I work for, I cannot choose the language we work in.  The situation is similar to how state works: you don't opt in or opt out of your state.  If you committed a crime, but think that opting out of the state that prosecutes the crime will save you from punishment--it will not work.

So, all I can, practically do falls into these two bins:

1. Tell you that you did a very crappy job, and hope that you will eventually understand why.  (Perusing the metaphor: appeal to the authorities).
2. Start my own project to replace what you did.  (Rebel).

And, again, just like with the state, the limited and necessarily shared resources would not allow me to choose (2).  Just like with the state, I cannot choose to start my own state simply because I don't like the one I live in.  I need a plot of land, a recognition from UN and so on...  Meaning that even if I wrote my own package manager, vastly superior to everything PyPA managed to come up with so far, I will ultimately fail at my larger goal, because the program like package manager is ultimately as useful as the package repository that stands behind it.  Egg and Wheel are laughably bad formats, well, bad ideas, not even formats, but if I want to write a better package manager, I need to either use those, or it will not be useful as a package manager...

**Comment by dstufft** [comment]
You can also just write your own PEP for your own format, describe why it's better, and get it accepted if it's actually better. Nobody here is married to the wheel format.

**Comment by gaborbernat** [comment]
@wvxvw the crust of thing is that almost no-one here is paid to make packaging better. So the nature of all packaging software is voluntary and constraint-driven. We all do it part of our free time, which is very limited (we have our full-time job just as well as you). As such our resources are probably just as limited as you described for yourself, and we need to optimize for the best options based on that. This does mean we often try to solve the problem for the most, not all. If your company really suffers because of this, feel free to convince them that it's really important to have a great solution for your problem, and go ahead and do it. Once you prove it with your use cases it will probably be easier to convince us that we should adopt this, instead of whatever we're using at the moment.

**Comment by njsmith** [comment]
Y'all, you're not required to spend a bunch of energy talking to someone after they've explicitly said they're not interested in working constructively. It's not healthy, and it rewards the abusive person at the expense of everyone else. Just block and move on.

**Comment by theacodes** [comment]
I am locking this conversation. It's so far beyond constructive. If there are actionable things that came out of this discussion, I encourage *PyPA members* to file bugs against the relevant project(s).

If you have questions or concerns about me locking this conversation, feel free to reach out to me at magicalgirl@google.com or to the packaging working group.
